tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts

```diff
@@ -295,6 +295,72 @@ export function buildOrReturnStateWithPinnedCopy(
   };
 }
 
+/**
+ * Returns updated cardToPinnedCopy, pinnedCardToOriginal, and cardMetadataMap.
+ * Under the same namespace, cardToPinnedCopy and pinnedCardToOriginal are
+ * routeful state and remained still, which might stil include the outdated
+ * mapping of removed experiments. This function handles the mappings.
+ * In the reducer nextCardmetadatamap is created from empty instead of
+ * preserving the previous map, which might also contain outdated mapping due to
+ * the same reason. To preserve the pinned card mapping we need to adding the
+ * pinned card mappng to cardmetadatamap.
+ */
+export function generateNextPinnedCardMappings(
+  cardToPinnedCopy: CardToPinnedCard,
+  pinnedCardToOriginal: PinnedCardToCard,
+  nextCardMetadataMap: CardMetadataMap,
+  nextCardList: CardId[]
+) {
+  const nextCardToPinnedCopy = new Map() as CardToPinnedCard;
+  for (const cardId of nextCardList) {
+    if (cardToPinnedCopy.has(cardId)) {
+      nextCardToPinnedCopy.set(cardId, cardToPinnedCopy.get(cardId)!);
+    }
+  }
+
+  const nextPinnedCardToOriginal = new Map(pinnedCardToOriginal);
+  // Removes previous pinned cards which are not in nextCardList; updated when
+  // experiments are removed
+  for (const [pinnedCardId, originalCardId] of pinnedCardToOriginal.entries()) {
+    if (!nextCardList.includes(originalCardId)) {
+      nextPinnedCardToOriginal.delete(pinnedCardId);
+    }
+  }
+
+  // Creates pinnedCardMetadataMap to preserve the mapping of pinned cards .
+  const pinnedCardMetadataMap = {} as CardMetadataMap;
+  for (const [
+    pinnedCardId,
+    originalCardId,
+  ] of nextPinnedCardToOriginal.entries()) {
+    if (nextCardToPinnedCopy.has(originalCardId)) {
+      pinnedCardMetadataMap[pinnedCardId] = nextCardMetadataMap[originalCardId];
+    }
+  }
+
+  return {
+    nextCardToPinnedCopy,
+    nextPinnedCardToOriginal,
+    pinnedCardMetadataMap,
+  };
+}
+
+/**
+ * Removes cards not in cardMetadataMap from cardStepIndex mapping.
+ */
+export function generateNextCardStepIndex(
+  cardStepIndex: CardStepIndexMap,
```

Comment:
For consistency with generateNextPinnedCardMappings function, I'd recommend renaming these arguments to previousCardStepIndex and nextCardMetadataMap.